### ISSUE TITLE
Character encoding problem importing bibtex

### ISSUE THREAD

**Issue by ar-jan** [leo-editor/leo-editor#381]
I'm not sure if this is related to the other unicode related problems in #212 and #147, but since it's happening on Windows and without using the clipboard it looks like a separate issue.

To reproduce: Enable the bibtex plugin; create a UTF-8 encoded .bib file containing e.g. an accented character like `é`. Import the file by creating an `@bibtex test.bib` node and double-clicking it.

Example:
```
@article{lastname_2017,
	author = {Lastname, Sébastien},
	year = {2017},
	title = {Article Title},
	journal = {My Journal},
	volume = {52},
	number = {3},
	pages = {23--48},
}
```
After importing, Sébastien becomes `SÃ©bastien` in Leo.

Any ideas?

--
Leo 5.4, build 20170120083547, Fri Jan 20 08:35:47 CST 2017
Git repo info: branch = master, commit = ecfd349e29a6
Python 3.5.2, PyQt version 5.6.2
Windows 7 AMD64 (build 6.1.7601) SP1

**Comment by edreamleo** [comment]
I'll have a look at this after I put out the [recent fires](https://groups.google.com/d/msg/leo-editor/tER1UGoxKkk/yRfWoDY6AwAJ).

**Comment by ar-jan** [comment]
Hm, so what seems to happen here is that Python defaults to using the system's codepage when working with files in text mode. On my Windows PC this is Windows-1252 (cp1252).

If I add "encoding='utf-8'" to the open() function of bibtex.py so it reads `bibFile = open(fn,'r', encoding='utf-8')`, it reads the UTF8-encoded file correctly.

I suppose the proper thing to do would be to always open files as binary, and try to guess the encoding, perhaps using something like BeautifulSoup's [UnicodeDammit](http://bazaar.launchpad.net/~leonardr/beautifulsoup/bs4/view/head:/bs4/dammit.py)?

**Comment by edreamleo** [comment]
@ar-jan: Leo is not going guess encodings, no matter how good the guesses usually are. I don't think a new setting is a great idea.  Instead, readBibTexFileIntoTree should honor Leo's `@encoding` directive, with the usual defaults, which are usually utf-8 :-)

Btw, `open(fn,'r', encoding='utf-8')` is Python 3 only.  [This page](http://stackoverflow.com/questions/10971033/backporting-python-3-openencoding-utf-8-to-python-2) discusses back ports to Python 2.

**Comment by edreamleo** [comment]
There is no BibTeX importer, so the import-file command will create an `@clean` node. I created the test file you described, and imported it.  I got:

```python
@ignore
@language bibtex
# -*- coding: utf-8 -*-

@article{lastname_2017,
	author = {Lastname, Sébastien},
	year = {2017},
	title = {Article Title},
	journal = {My Journal},
	volume = {52},
	number = {3},
	pages = {23--48},
}
```

As you can see, there are two problems. Besides the `@ignore` directive, the encoding line should be:

```python
@first # -*- coding: utf-8 -*-
```
Removing the `@ignore` and using `@first` causes Leo to round-trip the file properly.

But this is a strange hack.  Why should the *general* importer use the *python* encoding line? Clearly, something is not right in the general importer code. Furthermore, replacing `@first # -*- coding: utf-8 -*-` with `@encoding utf-8` writes byte hash (SÃ©bastien) back to the file.

In short, there appear to be several issues to be fixed here.

P. S. There are problems with the bibtex plugin. I'll fix them as part of this issue.

**Comment by ar-jan** [comment]
Yes, using Leo's default encoding sounds fine. Thanks for looking into this.

I hadn't actually used the import from the GUI, with that (and with `@auto test.bib`) it does use utf-8 for me too.

But, this just puts the bibtex content in the node. The docs in bibtex.py say `Creates a BibTex file from an  '@bibtex <filename>' tree`. And when I do this, the plugin does some further processing: every bibtex entry gets a Leo node, in this case with `@article lastname_2017` headline, and with contents like this:

```
author = {Lastname, Sébastien},
year = {2017},
title = {Article Title},
journal = {My Journal},
volume = {52},
number = {3},
pages = {23--48},
```

So the `@article{lastname_2017,` and closing `}` parts of the raw bibtex are moved to the outline structure.

The advantage would be the possibility of creating different .bib files from a master bibliography managed in Leo (with rearranging and organizing via Leo nodes).

**Comment by ar-jan** [comment]
Your latest commit fixes the encoding issue for me, when I import the test file via a `@bibtex test.bib` node.

Is this technically a Leo importer, or does the plugin do something else here?

**Comment by ar-jan** [comment]
Spoke to soon. It correctly reads the file, but, when using the plugin's write procedure by double-clicking the @bibtex node, it writes an ANSI-encoded .bib file (according to Notepad++). And there are also extra newlines added for every line: it now writes `CR` plus `CRLF` (before, only `CRLF`).

**Comment by edreamleo** [comment]
On Thu, Feb 23, 2017 at 10:08 AM, Arjan <notifications@github.com> wrote:

> Your latest commit fixes the encoding issue for me, when I import the test
> file via a @bibtex test.bib node.
>
> Is this technically a Leo importer, or does the plugin do something else
> here?
>
​The plugin is self-contained. It uses the new g.getEncoding(p, s)
function, which honors any @encoding directive in effect for p.  On
Windows, g.getEncoding uses locale.getpreferredencoding() if
s.encode('utf-8', 'strict') fails. This is apparently the recommended way
to get the Windows code page.

This should solve all reading problems, assuming that any @encoding
directives are telling the truth.

Edward

**Comment by edreamleo** [comment]
On Thu, Feb 23, 2017 at 10:30 AM, Arjan <notifications@github.com> wrote:

> Spoke to soon. It correctly reads the file, but, when using the plugin's
> write procedure by double-clicking the @bibtex <https://github.com/bibtex>
> node, it writes an ANSI-encoded .bib file (according to Notepad++). And
> there are also extra newlines added for every line: it now writes CR plus
> CRLF (before, only CRLF).
>
​Fixed, I think, at 845254e.

Edward

**Comment by ar-jan** [comment]
Thanks, I think it's working as intended. But just to be sure, because I'm not certain about which settings are are meant for which parts, or which take precedence, a few questions/clarifications:

I found the following in my global leoSettings.leo (and nothing in my personal settings):
Under "Basic setting nodes", a `@encoding` setting node is mentioned, which should be of the type headline = `@type-name setting-name = value`, but there is no default setting (only an empty `@enoding` node). Should I configure this in my personal settings, and which parts is it responsible for?

Then under "Unicode options", there's also these:
```
@string default_at_auto_file_encoding = utf-8
@string default_derived_file_encoding = utf-8
@string new_leo_file_encoding = utf-8
@string qt-default-encoding = UTF-8
```

When I create e.g. an `@file test.txt` file and put some content in it, I get a UTF-8 encoded .txt file by default, without changing anything about encodings. I suppose this is becuase of the above default_derived_file_encoding setting?

When I import the test.bib file with a `@bibtex test.bib` headline, importing works whether I put a `@encoding utf-8` directive in the node body, or not. But in order to write a utf-8 encoded .bib file via the plugin, I do need to put `@encoding utf-8` directive in the `@bibtex test.bib` node's body (otherwise it still writes an ANSI-encoded file, also at git rev d251a9b).

Is this the correct behavior? And does the `@encoding utf-8` directive apply only to the `@file` node where it's specified, or can I make this the default globally?

**Comment by edreamleo** [comment]
On Thu, Feb 23, 2017 at 12:15 PM, Arjan <notifications@github.com> wrote:

> Thanks, I think it's working as intended.
>
​Good to know.  I have just upped a few supposedly minor changes. Let me
know if there are problems.​

> But just to be sure, because I'm not certain about which settings are are
> meant for which parts, or which take precedence, a few
> questions/clarifications:
>
> I found the following in my global leoSettings.leo (and nothing in my
> personal settings):
> Under "Basic setting nodes", a @encoding setting node is mentioned, which
> should be of the type headline = @type-name setting-name = value, but
> there is no default setting (only an empty @enoding node)
>
Should I configure this in my personal settings, and which parts is it
> responsible for?
>
​No, I don't think so.​


> Then under "Unicode options", there's also these:
>
> @string default_at_auto_file_encoding = utf-8
> @string default_derived_file_encoding = utf-8
> @string new_leo_file_encoding = utf-8
> @string qt-default-encoding = UTF-8
>
> When I create e.g. an @file test.txt file and put some content in it, I
> get a UTF-8 encoded .txt file by default, without changing anything about
> encodings. I suppose this is becuase of the above
> default_derived_file_encoding setting?
>
​Probably ;-)  As you can see, utf-8 is the default everywhere.​

> When I import the test.bib file with a @bibtex test.bib headline,
> importing works whether I put a @encoding utf-8 directive in the node
> body, or not. But in order to write a utf-8 encoded .bib file via the
> plugin, I do need to put @encoding utf-8 directive in the @bibtex test.bib
> node's body (otherwise it still writes an ANSI-encoded file, also at git
> rev d251a9b
> <https://github.com/leo-editor/leo-editor/commit/d251a9b4265f26574bc7f28af1c7e71e27055a24>
> ).
>
Is this the correct behavior?
>
​No. Please test with ​
​e87f716. The default encoding is now utf-8.​
​

> And does the @encoding utf-8 directive apply only to the @file node where
> it's specified, or can I make this the default globally?
>
​The @encoding directive applies to a node and its descendants. The bibtex
plugin's write logic now does:

    encoding=g.getEncoding(root)
    with open(fn, 'wb') as f:
        s = ''.join(strings + entries)
        f.write(g.toEncodedString(s,encoding=encoding))

which will, in fact, default to utf-8 if g.getEncoding(root) is None.

The defaults you mention above do not apply to the bibtex plugin.  Perhaps
they should, but I think the latest code will work for just about everybody.

Edward

**Comment by ar-jan** [comment]
Thanks! I'll test tomorrow when I'm at a Windows machine again.

**Comment by edreamleo** [comment]
Rev c8d2f82 disables the dubious Windows-only code in `g.getEncodingAt`. Please let me know if it works for you. I am going to close this issue for now.  Please reopen if needed.

The disabled code uses `locale.getpreferredencoding()` if the string is not unicode. I would much prefer to do without such special cases.

**Comment by ar-jan** [comment]
All is working well now (at current master). Thanks!

**Comment by edreamleo** [comment]
On Fri, Feb 24, 2017 at 7:23 AM, Arjan <notifications@github.com> wrote:

> All is working well now (at current master). Thanks!
>
​Excellent. Thanks for the confirmation.

I am going to leave g.getEncodingAt as it is, *without *the Windows-only
special case code for locale.getpreferredencoding().

Edward
